### ISSUE TITLE
plugin/health: allow per server health

### ISSUE THREAD

**Issue by fturib** [coredns/coredns#1686]
## Description

When trying to run CoreDNS with this Corefile 👍 

```
.:5002 francois.com:5002 {
    errors
    log
    health localhost:8080
    prometheus localhost:8075
	proxy . 8.8.8.8 {
		protocol dns force_tcp
	}
	reload 2s
}
```

I get the error:
`2018/04/17 17:11:54 listen tcp 127.0.0.1:8080: bind: address already in use`

## Invstigation

the reason is that the Setup for any plugin is called once per each head zone/port (listener). In current use case these are 2 : ".:5002" and "francois.com:5002".

The Plugin create one HEALTH structure for each call, and then will try to open one HTTP Listener for each HEALTH structure.

## Resolution in progress ...

**Comment by miekg** [comment]
yes, health needs to be the same thing as the prometheus plugin.
This is currently non-obvious but mentioned in the README;
~~~
If you have multiple Server Block and need to export health for each of the plugins, you must run health endpoints on different ports:
~~~

**Comment by miekg** [comment]
Coming back to this, while trying to fix it, your corefile is similar to:
~~~
.:5002 {
    errors
    log
    health localhost:8080
    prometheus localhost:8075
	proxy . 8.8.8.8 {
		protocol dns force_tcp
	}
	reload 2s
}

francois.com:5002 {
    errors
    log
    health localhost:8080
    prometheus localhost:8075
	proxy . 8.8.8.8 {
		protocol dns force_tcp
	}
	reload 2s
}
~~~

This is explicitly called out in the README:
> If you have multiple Server Blocks and need to export health for each of the plugins, you must run
> health endpoints on different ports:
> 
> ~~~ corefile
>com {
>   whoami
>   health :8080
>}
>
> net {
>    erratic
>    health :8081
>}
>~~~

There is no way (as of yet) to limit this to a specific server, so this is WAI

**Comment by miekg** [comment]
If we want to fix this we should do per-server health and have a process health scraping all those per-server health points.
In fact, I see no other way to support this.

**Comment by fturib** [comment]
The point is to dissociate the HTTP Servers from the Health object itself.
- one side : HTTP Server : we want one per address specified
- other side : Health object : Only collecting the Health info for all the Handler chains that are created from this Plugin. 

That was implemented that way in this PR : #1688. (but using also variables per instance ...)

### Simple way - Global var, but not RELOAD compatible
The whole point is to have a "global" variable in the right scope that Plugins and/or Handler can use to save some shared information (the Healther HTTP Servers are shared information of all Health Plugins).

The issue is that right now, we can use only the global variable (like the Address for Metrics). 
But these global variables are incompatible with RELOAD

NOTE: it is not RELOAD compatible because, the during a reload, the global var is altered at PLUGIN setup time .. and if the Corefile is wrong, there is no rollbak : CoreDNS keep running on the former instance, but with data that was altered ...


### a safer way for RELOAD - Allow Variable per instance (in the dnsContext)
=> we would need global variable within the scope of the current Caddy instance.
A way to do it, is to have ability to save and get variables in the dnsContext : we have one per instance of Caddy.

Known issue: sometime to avoid overlapping process level resource, you need to KNOW the status in the former instance. That means you would need somehow to be able to query the variable of the former instance. Feasible but had some complexity.

see : #1688.


### the right way (IMO) : predefine scope of variables that can be used by Plugin depending their needs.
It is right now confusing for the developer what is the scope of the current Plugin : it is per server bloc, per listener (consider all BIND addresses in the same scope, event though not on the same listener)

Metrics would need : per HTTP listener address
Reload would need : per caddy instance
Health would need : per HTTP listener address
Rewrite would need : per Server bloc (same for all listener is good enough, after it is duplication for nothing)
etc ...

I guess we should propose a mechanism for the Plugins to define their scope of sharing the same Plugin object, and ability to save data at some pre-defined or user-defined level of scope.
